Add status field to TransportServer resource

soneillf5 opened the pull request:
### Proposed changes
This change adds the status field to the TransportServer resource.
This enables a user to get the status of the TransportServer i.e.
```
$ kubectl get ts
NAME      STATE   REASON           IP    PORTS   AGE
dns-tcp   Valid   AddedOrUpdated                 54m
```

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

pleshakov reviewed (commented):
Hi @soneillf5 

Please see my feedback.

With TransportServer, it is a bit trickier than with VirtualServer: VirtualServer can have one or two ports. For example it can have port `80` or ports `80` and `443`. However, with TransportServer, we only have one port, which corresponds to the port of the listener.

For TCP/UDP TransportServer , it only makes sense to report IP and port when:
* external-status-address is set in the ConfigMap (we assume that address exposes all ports)
* the type LoadBalancer service, configured for KIC and used via -external-service, configures  a port and protocol that matches the port and protocol of the listener used in the TransportServer. So if the Services exposes port 53, we report the service public IPs and port 53 for a TransportServer that has a listener that uses port 53. Otherwise, if the port is not configured in the Service, it will not be accessible via the allocated cloud LoadBalancer.

For TLS Passthrough TransportServer, which uses port `443`, it makes sense to report IP and port in the status when:
* external-status-address is set in the ConfigMap
* there is a  type LoadBalancer services, configured for KIC and used via -external-service. The services we provide in the installation manifests, helm and operator, have ports 80 and 443 configured. 
* IngressLink is configured https://github.com/nginxinc/kubernetes-ingress/blob/master/docs-web/integration-with-cis.md (IngressLink supports ports 80 and 443)

soneillf5 commented on internal/k8s/configuration.go:
```diff
@@ -764,8 +764,9 @@ func (c *Configuration) GetResources() []Resource {
-	VirtualServers bool
+	Ingresses        bool
+	VirtualServers   bool
+	TransportServers bool
```
controller.go L1681

pleshakov replied:
ok, it is used for write, but not for read -- GetResourcesWithFilter doesn't take in into consideration and doesn't return any  TransportServers

soneillf5 commented on internal/k8s/controller.go:
```diff
@@ -1454,25 +1469,35 @@ func (lbc *LoadBalancerController) updateRegularIngressStatusAndEvents(ingConfig
 		eventType = api_v1.EventTypeWarning
 		eventTitle = "AddedOrUpdatedWithWarning"
 		eventWarningMessage = fmt.Sprintf("with warning(s): %s", formatWarningMessages(tsConfig.Warnings))
+		state = conf_v1.StateInvalid
```
@LorcanMcVeigh ^ ?

LorcanMcVeigh replied:
I think it's ok to leave the state as either `Valid` or `Invalid` because having the state be `Warning` doesn't tell whether the resource is valid or not. But leaving the state as `Valid` or `Invalid` and having the reason next to it with the message `AddedOrUpdatedWithWarning` would make things a lot clearer 
```$ kubectl get transportserver
  NAME      STATE   REASON                      IP    PORTS   AGE
  dns-tcp   Valid   AddedOrUpdatedWithWarning                 47m
  ```
  This is a lot clearer, in relation to the state of the TransportServer than 
  ```$ kubectl get transportserver
  NAME      STATE     REASON                      IP    PORTS   AGE
  dns-tcp   Warning   AddedOrUpdatedWithWarning                 47m
  ```

LorcanMcVeigh replied:
I think this is something we can change for VirtualServers and VirtualServerRoutes as well, but at another time

soneillf5 replied:
I've compromised and gone with StateWarning. I'm still not sure it's correct though.

soneillf5 commented on pkg/apis/configuration/v1alpha1/types.go:
```diff
@@ -117,6 +124,20 @@ type Action struct {
+// ExternalEndpoint defines the IP and ports used to connect to this resource.
+type ExternalEndpoint struct {
+	IP    string `json:"ip"`
+	Ports string `json:"ports"`
```
I've updated it to port but then it doesn't map cleanly to the existing endpoint code. 
So fix this, I've also added a externalListenerEndpoints ( not sure about the name ) to capture the difference. 
We now also need more code and complexity to handle these.

Are you sure this is correct?

pleshakov replied:
externalListenerEndpoints 👍 
If we only report for TLS Passthrough, then `externalTLSListenerEndpoints` would also make sense to me

soneillf5 commented on internal/k8s/controller.go:
```diff
@@ -865,19 +866,30 @@ func (lbc *LoadBalancerController) syncTransportServer(task task) {
 	lbc.processProblems(problems)
 }
 
+func (lbc *LoadBalancerController) addOrUpdateTransportServer(key string, ts *conf_v1alpha1.TransportServer) ([]ResourceChange, []ConfigurationProblem) {
```
updated

soneillf5 commented on internal/k8s/utils.go:
```diff
@@ -199,3 +199,8 @@ func GetK8sVersion(client kubernetes.Interface) (v *version.Version, err error)
 }
+
+// storeToTransportServerLister makes a Store that lists TransportServers
+type storeToTransportServerLister struct {
```
good catch, removed

soneillf5 commented on internal/k8s/status_test.go:
```diff
@@ -15,6 +17,189 @@ import (
+	}
+	updatedTs, _ := fakeClient.K8sV1alpha1().TransportServers(ts.Namespace).Get(context.TODO(), ts.Name, meta_v1.GetOptions{})
+
+	if updatedTs.Status.State != "after status" {
```
Much cleaner, thanks for the suggestion. I've updated it.

soneillf5 commented on internal/k8s/status_test.go:
```diff
@@ -15,6 +17,189 @@ import (
+			}})
+
+	tsLister := storeToTransportServerLister{}
+	tsLister.Store, _ = cache.NewInformer(
```
yup, much better. Updated.

soneillf5 commented on internal/k8s/status_test.go:
```diff
@@ -15,6 +17,189 @@ import (
+
+	updatedTs, _ := fakeClient.K8sV1alpha1().TransportServers(ts.Namespace).Get(context.TODO(), ts.Name, meta_v1.GetOptions{})
+	if updatedTs != nil {
+		t.Errorf("expected TransportServer Store would be empty as provided TransportServer was not found")
```
done

soneillf5 commented on internal/k8s/controller.go:
```diff
@@ -1652,6 +1677,15 @@ func (lbc *LoadBalancerController) syncService(task task) {
+
+			glog.V(3).Infof("Updating status for %v TransportServers", len(transportServers))
+
+			err = lbc.statusUpdater.UpdateExternalEndpointsForResources(transportServers)
```
I've added some code which I think will handle it now.

soneillf5 commented:
Hi @pleshakov, thank for the review. I've addressed most of the comments.
 
The outstanding issue is your comment about reporting the external IP/Port. 
You state 'it only makes sense to report the IP/host' under certain conditions.
My understanding is that the output of `k get ts` depends on the CRD `k8s.nginx.org_transportservers.yaml` which we don't change at runtime? If that's the case, how do we change what we report?

pleshakov commented on pkg/apis/configuration/v1alpha1/types.go:
```diff
@@ -50,13 +50,20 @@ type GlobalConfigurationList struct {
+// +kubebuilder:printcolumn:name="State",type=string,JSONPath=`.status.state`,description="Current state of the VirtualServer. If the resource has a valid status, it means it has been validated and accepted by the Ingress Controller."
+// +kubebuilder:printcolumn:name="Reason",type=string,JSONPath=`.status.reason`
+// +kubebuilder:printcolumn:name="IP",type=string,JSONPath=`.status.externalEndpoints[*].ip`
+// +kubebuilder:printcolumn:name="Ports",type=string,JSONPath=`.status.externalEndpoints[*].ports`
```
```suggestion
// +kubebuilder:printcolumn:name="Ports",type=string,JSONPath=`.status.externalEndpoints[*].port`
```

pleshakov commented on pkg/apis/configuration/v1alpha1/types.go:
```diff
@@ -50,13 +50,20 @@ type GlobalConfigurationList struct {
 // +kubebuilder:validation:Optional
 // +kubebuilder:resource:shortName=ts
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="State",type=string,JSONPath=`.status.state`,description="Current state of the VirtualServer. If the resource has a valid status, it means it has been validated and accepted by the Ingress Controller."
```
VirtualSever -> TransportServer?

pleshakov commented on pkg/apis/configuration/v1alpha1/types.go:
```diff
@@ -117,6 +124,20 @@ type Action struct {
+	ExternalEndpoints []ExternalEndpoint `json:"externalEndpoints,omitempty"`
+}
+
+// ExternalEndpoint defines the IP and ports used to connect to this resource.
```
```suggestion
// ExternalEndpoint defines the IP and port used to connect to this resource.
```

pleshakov commented on internal/k8s/status_test.go:
```diff
@@ -15,6 +18,171 @@ import (
+				*ts,
+			}})
+
+	tsLister, _ := cache.NewInformer(
```
would it make sense to also simplify here and TestUpdateTransportServerStatusMissingTransportServer ? 
```
tsLister := cache.NewStore(cache.DeletionHandlingMetaNamespaceKeyFunc)
```

pleshakov commented on internal/k8s/status.go:
```diff
@@ -560,6 +652,21 @@ func (su *statusUpdater) generateExternalEndpointsFromStatus(status []api_v1.Loa
+	for _, lb := range status {
+		port := su.externalServicePorts
+		if su.bigIPPorts != "" {
+			port = su.bigIPPorts
```
in case of bigIPPorts, `u.bigIPPorts` will have `[80,443]` here. We need to choose just `443` (assuming we only report IP for Passthrough TransportServers)

pleshakov commented on internal/k8s/status.go:
```diff
@@ -560,6 +652,21 @@ func (su *statusUpdater) generateExternalEndpointsFromStatus(status []api_v1.Loa
+func (su *statusUpdater) generateExternalListenerEndpointsFromStatus(status []api_v1.LoadBalancerIngress) []conf_v1alpha1.ExternalEndpoint {
+	var externalEndpoints []conf_v1alpha1.ExternalEndpoint
+	for _, lb := range status {
+		port := su.externalServicePorts
```
Here we need to get the port that matches `https` in the LoadBalancer service -- https://github.com/nginxinc/kubernetes-ingress/blob/master/deployments/service/loadbalancer.yaml#L14 (assuming we only report IP for Passthrough TransportServers)

Could we do something like we do in `getExternalServicePorts` ? go through the ports of the svc and get the one that matches `https`?   maybe that could be done in a separate function so that the port value is passed to `generateExternalListenerEndpointsFromStatus` as an argument?

soneillf5 commented:
The `IP` and `Port` fields are a lot more nuanced and involved, causing a lot of rework on this PR.

I've removed the two fields and logic required for them. If they are still required, we can add them in a new PR.

pleshakov commented on pkg/apis/configuration/v1alpha1/types.go:
```diff
@@ -50,13 +50,20 @@ type GlobalConfigurationList struct {
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="State",type=string,JSONPath=`.status.state`,description="Current state of the VirtualServer. If the resource has a valid status, it means it has been validated and accepted by the Ingress Controller."
+// +kubebuilder:printcolumn:name="Reason",type=string,JSONPath=`.status.reason`
+// +kubebuilder:printcolumn:name="IP",type=string,JSONPath=`.status.externalEndpoints[*].ip`
```
shall we remove the IP and port print columns for now as well?

(crds will need to be re-generated after that)

soneillf5 replied:
done

pleshakov commented on internal/k8s/controller.go:
```diff
@@ -1454,25 +1456,35 @@ func (lbc *LoadBalancerController) updateRegularIngressStatusAndEvents(ingConfig
 		eventType = api_v1.EventTypeWarning
 		eventTitle = "AddedOrUpdatedWithWarning"
 		eventWarningMessage = fmt.Sprintf("with warning(s): %s", formatWarningMessages(tsConfig.Warnings))
+		state = conf_v1.StateWarning
```
could you update the doc and mention that Warning is one of the possible values?

soneillf5 replied:
done

pleshakov commented on internal/k8s/controller.go:
```diff
@@ -1652,6 +1664,15 @@ func (lbc *LoadBalancerController) syncService(task task) {
 				glog.V(3).Infof("error updating VirtualServer/VirtualServerRoute status in syncService: %v", err)
 			}
+
+			transportServers := lbc.configuration.GetResourcesWithFilter(resourceFilter{TransportServers: true})
```
I think this part could be removed as well, considering IP will not be reported.
because in syncService we can only trigger status update related to the update in the IP

Also, I mentioned in this comment https://github.com/nginxinc/kubernetes-ingress/pull/1425/files#r587905227 about resourceFilter{TransportServers: true}, because GetResourcesWithFilter doesn't read that value (more over, it doesn't return TransportServers at all). If you remove this part, I suggest you remove that field in the filter as well

soneillf5 replied:
done

soneillf5 commented:
@pleshakov updated and ready to merge

pleshakov reviewed (approved):
@soneillf5 lgtm

(but I can't find this though https://github.com/nginxinc/kubernetes-ingress/pull/1425#discussion_r589382217 )
